src/modules/bootloader/main.py

```diff
@@ -637,25 +637,47 @@ def run_grub_install(fw_type, partitions, efi_directory):
                                    "--bootloader-id=" + efi_bootloader_id,
                                    "--force"])
     else:
-        assert efi_directory is None
-        if libcalamares.globalstorage.value("bootLoader") is None:
-            return
+        try:
+            install_hybrid_grub = libcalamares.job.configuration["installHybridGRUB"]
+        except KeyError:
+            install_hybrid_grub = False
 
-        boot_loader = libcalamares.globalstorage.value("bootLoader")
-        if boot_loader["installPath"] is None:
-            return
+        if efi_directory is None:
+            if libcalamares.globalstorage.value("bootLoader") is None:
+                return
+
+            boot_loader = libcalamares.globalstorage.value("bootLoader")
+            boot_loader_install_path = boot_loader["installPath"]
+            if boot_loader_install_path is None:
+                return
+        else:
+            if not install_hybrid_grub:
+                return
+
+            # We're actually on an EFI system even though fw_type is "bios".
+            # We need to find the proper drive to install to based on which
+            # drive the ESP is located on.
+            installation_root_path = libcalamares.globalstorage.value("rootMountPoint")
+            install_efi_directory = installation_root_path + efi_directory
+            find_esp_disk_command = "echo \"/dev/$(lsblk -o PKNAME $(df --output=source" \
+                + install_efi_directory \
+                + " | tail -n1) | tail -n1)\""
```

Comment:
Would this work(and be simpler)?  `f"findmnt -no source {install_efi_directory}"`

Also, try to use f strings when it can improve readability.  When the module was written it wasn't supported but for new code it is available now.

That being said, I am not sure this should be really be happening here.  It seems like this should come from the partition module.

Reply:
`findmnt` doesn't quite work because it gives me the partition device file, not the disk device file (which is what `grub-install` wants when doing a BIOS-style installation). It could probably replace the `df` stuff in the innermost subshell, but the `lsblk` and `echo` will still be needed if this kind of logic is used.

Reply:
Is that not in global storage already in `partitions`?

Reply:
It does not seem to be based on some debugging. On BIOS systems, it's in the bootLoader key, but on EFI systems it's not there at all. On the VM I just tested this on, there's some strange "empty" device (`{'claimed': False, 'device': '', 'features': {}, 'fs': 'unknown', 'fsName': 'unknown', 'mountPoint': '', 'partattrs': 0, 'partlabel': '', 'parttype': '', 'partuuid': '', 'uuid': ''}`), then there's four devices, one for each of /dev/vda1 through /dev/vda4. /dev/vda itself isn't in the list.

I don't want to just transform the partition name to a disk name since it's not guaranteed the two will have a clear correlation  - we see a minor version of this with /dev/sda and /dev/sda1 vs. /dev/mmcblk0 and /dev/mmcblk0p1.